bulk-model-sync-lib/src/commonMain/kotlin/org/modelix/model/sync/bulk/ModelImporter.kt

```diff
@@ -121,52 +122,127 @@ class ModelImporter(private val root: INode, private val continueOnError: Boolea
+            syncChildRoleOrdered(expectedNodes, existingNodes, progressReporter, node, role)
+        } else {
+            syncChildRoleUnordered(expectedNodes, node, role, progressReporter)
+        }
+    }
+
+    private fun syncChildRoleOrdered(
+        expectedNodes: List<NodeData>,
+        existingNodes: List<INode>,
+        progressReporter: ProgressReporter,
+        node: INode,
+        role: String?,
+    ) {
+        // optimization for when there is no change in the child list
+        // size check first to avoid querying the original ID
+        if (expectedNodes.size == existingNodes.size && expectedNodes.map { it.originalId() } == existingNodes.map { it.originalId() }) {
+            existingNodes.zip(expectedNodes).forEach { syncNode(it.first, it.second, progressReporter) }
+            return
+        }
+
+        expectedNodes.forEachIndexed { index, expected ->
+            val nodeAtIndex = node.getChildren(role).toList().getOrNull(index)
+            val expectedId = checkNotNull(expected.originalId()) { "Specified node '$expected' has no id" }
+            val expectedConcept = expected.concept?.let { s -> ConceptReference(s) }
+            val childNode = if (nodeAtIndex?.originalId() != expectedId) {
+                val existingNode = originalIdToExisting[expectedId]
+                if (existingNode == null) {
+                    val newChild = node.addNewChild(role, index, expectedConcept)
+                    newChild.setPropertyValue(NodeData.idPropertyKey, expectedId)
+                    originalIdToExisting[expectedId] = newChild
+                    newChild
                 } else {
-                    nodeAtIndex
+                    // The existing child node is not only moved to a new index,
+                    // it is potentially moved to a new parent and role.
+                    node.moveChild(role, index, existingNode)
+                    // If the old parent and old role synchronized before the move operation,
+                    // the existing child node would have been marked as to be deleted.
+                    // Now that it is used, it should not be deleted.
+                    nodesToRemove.remove(existingNode)
+                    existingNode
                 }
-                check(childNode.getConceptReference() == expectedConcept) { "Unexpected concept change" }
+            } else {
+                nodeAtIndex
+            }
+            check(childNode.getConceptReference() == expectedConcept) { "Unexpected concept change" }
+
+            syncNode(childNode, expected, progressReporter)
+        }
+
+        // At this point, all n expected children for this role are created and correctly sorted.
+        // This means the first n children are correct.
+        // Any child beyond that is an unexpected child has to be marked for removal.
+        nodesToRemove += node.getChildren(role).drop(expectedNodes.size)
+    }
+
+    private fun syncChildRoleUnordered(
```

Comment:
This method has a large overlap with syncChildRoleOrdered. Wouldn't it make more sense to only factor out the index computation into two distinct paths. Right now, property assignment etc is duplicated. That means we have to remember to fix both paths in case there is a bug etc.

Reply:
That is an interesting thought.
I can probably reuse the old algorithm as is, if I manipulate the provided expected order of children to match the order of the existing children for roles that have no significant order.
I will give it a try and see if this works.
(Luckily, I have automated tests to validate such a refactoring 😁)

Reply:
Done. I found a solution that has less duplication, but is less straightforward.